livre/lib/cache/TextureObject.cpp

```diff
@@ -35,213 +36,188 @@ namespace livre
+        // LA (luminance and alpha, common for fonts): width * height * type size bytes.
+        // RGB (color, no alpha): width * height * type size bytes.
+        // RGBA (color with alpha): width * height * type size bytes.
+
+        uint32_t elementSize = 0;
+        switch( _dataCache.getTextureType( ))
         {
-            case 1:
-                format = GL_RED;
+            case GL_UNSIGNED_BYTE:
+                elementSize = sizeof( char );
                 break;
-            case 3:
-                format = GL_RGB;
+            case GL_FLOAT:
+                elementSize = sizeof( float );
                 break;
-            default:
-                format = GL_RED;
+            case GL_UNSIGNED_SHORT:
+                elementSize = sizeof( short );
                 break;
         }
 
-        TexturePoolPtr pool = textureCachePtr_->getTexturePool( maxBlockSize,
-                                                                format,
-                                                                gpuDataType );
-        textureState_->texturePoolPtr = pool;
+        const Vector3ui& textureSize =
+                _dataSource.getVolumeInformation().maximumBlockSize;
+        return textureSize.product() * elementSize;
     }
 
-    initialize_( );
-
-    textureState_->texturePoolPtr->generateTexture( textureState_ );
-    LBASSERT( textureState_->textureId );
-    loadTextureToGPU_( );
-    lodTextureData_.reset( TextureDataObject::getEmptyPtr() );
-    return true;
-}
+    bool loadTextureToGPU() const
+    {
+        const NodeId nodeId( _textureObject.getId( ));
+        const LODNode& lodNode = _dataSource.getNode( nodeId );
+
+    #ifdef LIVRE_DEBUG_RENDERING
+        std::cout << "Upload "  << nodeId.getLevel() << ' '
+                  << lodNode.getRelativePosition() << " to "
+                  << _textureState->textureId << std::endl;
+    #endif
+        const Vector3ui& overlap = _dataSource.getVolumeInformation().overlap;
+
+        const Vector3ui& voxSizeVec =
+                lodNode.getBlockSize() + overlap * 2;
+
+        TextureDataObjectPtr rawData =
+                boost::static_pointer_cast< TextureDataObject >(
+                    _dataCache.get( _textureObject.getId( )));
+
+        _textureState->bind( );
+        glTexSubImage3D( GL_TEXTURE_3D, 0, 0, 0, 0,
+                         voxSizeVec[0], voxSizeVec[1], voxSizeVec[2],
+                         _texturePool.getFormat() ,
+                         _texturePool.getGPUDataType(),
+                         rawData->getDataPtr( ));
+
+        // Something went wrong with loading the data
+        // TODO: Log message
+        const GLenum glErr = glGetError();
+        if ( glErr != GL_NO_ERROR )
+        {
+            LBERROR << "Error loading the texture into GPU, error number : "  << glErr << std::endl;
+            return false;
+        }
 
-void TextureObject::unload_( )
-{
-    LBASSERT( isValid_() );
+        glFinish( );
```

Comment:
This was what I have found in the past ( 2013 ): http://higherorderfun.com/blog/2011/05/26/multi-thread-opengl-texture-loading . It seemed like a driver issue so I did not spend time. If you guide me where you are looking at the spec, it would be helpful.